### PR TITLE
03622 wipe token precompile

### LINKED ISSUE #3622: HIP 514: Add token management functions for wipe functions
### Problem

The precompile flow is missing support for the wipe flow

### Solution

Add support for the wipe flow

- `wipeTokenAccount(address, address, uint32)`
- `wipeTokenAccount(address, address, [uint64])`

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by tdermendzhievv**
**Description**:
This PR introduces wipeTokenAccount precompile for fungible and non fungible token.

**Related issue(s)**:  

Fixes #3622 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/WipeWrapper.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import static com.hederahashgraph.api.proto.java.TokenType.FUNGIBLE_COMMON;
+import static com.hederahashgraph.api.proto.java.TokenType.NON_FUNGIBLE_UNIQUE;
+
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.TokenID;
+import com.hederahashgraph.api.proto.java.TokenType;
+import java.util.Collections;
+import java.util.List;
+
+public record WipeWrapper(TokenID token, AccountID account, long amount, List<Long> serialNos) {
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/WipeWrapper.java:27]: nit:
```suggestion
public record WipeWrapper(TokenID token, AccountID account, long amount, List<Long> serialNumbers) {
```

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/WipeWrapper.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    private static final long NONFUNGIBLE_WIPE_AMOUNT = -1;
+    private static final List<Long> FUNGIBLE_WIPE_SERIAL_NOS = Collections.emptyList();
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/WipeWrapper.java:30]: nit:
```suggestion
    private static final List<Long> FUNGIBLE_WIPE_SERIAL_NUMBERS = Collections.emptyList();
```

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/WipeWrapper.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        return new WipeWrapper(token, account, NONFUNGIBLE_WIPE_AMOUNT, serialNos);
+    }
+
+    public static WipeWrapper forFungible(
+            final TokenID token, final AccountID account, final long amount) {
+        return new WipeWrapper(token, account, amount, FUNGIBLE_WIPE_SERIAL_NOS);
+    }
+
+    public TokenType type() {
+        return (amount == NONFUNGIBLE_WIPE_AMOUNT) ? NON_FUNGIBLE_UNIQUE : FUNGIBLE_COMMON;
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/WipeWrapper.java:43]: Constructor is public so there's no assurance the amount passed in for non fungible case is exactly -1 is there?
Maybe safer to check for both non-empty serial list with null or <= 0 amount for `NON_FUNGIBLE_UNIQUE` case, then empty serial list with > 0 amount for `FUNGIBLE_COMMON` and a throw otherwise.
What do you think?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractWipePrecompile.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                syntheticTxnFactory,
+                infrastructureFactory,
+                pricingUtils);
+        this.aliases = aliases;
+        this.sigsVerifier = sigsVerifier;
+    }
+
+    @Override
+    public TransactionBody.Builder body(Bytes input, UnaryOperator<byte[]> aliasResolver) {
+        return null;
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractWipePrecompile.java:72]: Q: Why override just to set null? Why not just leave method out since extending class will implement?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractWipePrecompile.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @Override
+    public long getMinimumFeeInTinybars(Timestamp consensusTime) {
+        return 0;
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractWipePrecompile.java:77]: Q: Why override just to set 0? Why not just leave method out since extending class will implement?

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Lots of changes in here that are just from formatting.
Is a rebase needed to help readbaility?

Also please update the PR comments section with the intended changes.
This will help reviewers better understand the intended changes in what you're trying to achieve.

1st pass, will circle back again

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java [hunk 6]
```diff
@@ -1,119 +1,133 @@
...
+    public static final int ABI_ID_ERC_GET_APPROVED = 0x081812fc;
+    // isApprovedForAll(address owner, address operator)
+    public static final int ABI_ID_ERC_IS_APPROVED_FOR_ALL = 0xe985e9c5;
+    // ownerOf(uint256 tokenId)
+    public static final int ABI_ID_ERC_OWNER_OF_NFT = 0x6352211e;
+    // tokenURI(uint256 tokenId)
+    public static final int ABI_ID_ERC_TOKEN_URI_NFT = 0xc87b56dd;
+    // wipeTokenAccount(address, address, uint32)
+    public static final int ABI_WIPE_TOKEN_ACCOUNT_FUNGIBLE = 0x9790686d;
+    // wipeTokenAccount(address, address, int64[])
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java:102]: Update signature based on the new method name defined in HIP-514 -> wipeTokenAccountNFT

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 7]
```diff
@@ -38,765 +35,988 @@
...
+
+    private static final Function ERC_TRANSFER_FUNCTION =
+            new Function("transfer(address,uint256)", BOOL_OUTPUT);
+    private static final Bytes ERC_TRANSFER_SELECTOR = Bytes.wrap(ERC_TRANSFER_FUNCTION.selector());
+    private static final ABIType<Tuple> ERC_TRANSFER_DECODER =
+            TypeFactory.create("(bytes32,uint256)");
+
+    private static final Function WIPE_TOKEN_ACCOUNT_FUNCTION =
+            new Function("wipeTokenAccount(address,address,uint32)", INT_OUTPUT);
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:168]: nit: remove white line, so that WIPE_TOKEN variables are structured together

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 8]
```diff
@@ -38,765 +35,988 @@
...
+    private static final Bytes WIPE_TOKEN_ACCOUNT_SELECTOR =
+            Bytes.wrap(WIPE_TOKEN_ACCOUNT_FUNCTION.selector());
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:171]: nit: remove white line, so that WIPE_TOKEN variables are structured together

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 9]
```diff
@@ -38,765 +35,988 @@
...
+    private static final ABIType<Tuple> WIPE_TOKEN_ACCOUNT_DECODER =
+            TypeFactory.create("(bytes32,bytes32,uint32)");
+
+    private static final Function WIPE_TOKEN_ACCOUNT_NFT_FUNCTION =
+            new Function("wipeTokenAccountNFT(address,address,int64[])", INT_OUTPUT);
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:177]: nit: remove white line, so that WIPE_TOKEN_NFT variables are structured together

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 10]
```diff
@@ -38,765 +35,988 @@
...
+    private static final Bytes WIPE_TOKEN_ACCOUNT_NFT_SELECTOR =
+            Bytes.wrap(WIPE_TOKEN_ACCOUNT_NFT_FUNCTION.selector());
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:180]: nit: remove white line, so that WIPE_TOKEN_NFT variables are structured together

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractWipePrecompile.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        Objects.requireNonNull(wipeOp);
+
+        /* --- Check required signatures --- */
+        final var tokenId = Id.fromGrpcToken(wipeOp.token());
+        final var accountId = Id.fromGrpcAccount(wipeOp.account());
+        final var hasRequiredSigs =
+                KeyActivationUtils.validateKey(
+                        frame,
+                        accountId.asEvmAddress(),
+                        sigsVerifier::hasActiveKey,
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractWipePrecompile.java:77]: Introduce sigsVerifier::hasWipeKey

File: hedera-node/src/test/java/com/hedera/services/store/contracts/precompile/WipeNonFungiblePrecompileTest.java [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+        subject.prepareFields(frame);
+        subject.prepareComputation(pretendArguments, а -> а);
+        subject.getPrecompile().getGasRequirement(TEST_CONSENSUS_TIME);
+        subject.computeInternal(frame);
+
+        verify(wrappedLedgers, never()).commit();
+    }
+
+    @Test
+    void gasRequirementReturnsCorrectValueForBurnToken() {
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/test/java/com/hedera/services/store/contracts/precompile/WipeNonFungiblePrecompileTest.java:290]: Did you mean `gasRequirementReturnsCorrectValueForWipeToken` ?

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+                        contractCreate(WIPE_CONTRACT),
+                        tokenAssociate(ACCOUNT, VANILLA_TOKEN),
+                        cryptoTransfer(moving(500, VANILLA_TOKEN).between(TOKEN_TREASURY, ACCOUNT)))
+                .when(
+                        withOpContext(
+                                (spec, opLog) ->
+                                        allRunFor(
+                                                spec,
+                                                contractCreate(WIPE_CONTRACT)
+                                                        .bytecode(WIPE_CONTRACT),
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java:98]: It seems you create `WIPE_CONTRACT` twice. Also `bytecode` is not needed anymore to be passed explicitly.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+                                movingUnique(VANILLA_TOKEN, 1L, 2L)
+                                        .between(TOKEN_TREASURY, ACCOUNT)))
+                .when(
+                        withOpContext(
+                                (spec, opLog) -> {
+                                    final var serialNumbers = new ArrayList<>();
+                                    serialNumbers.add(1L);
+                                    allRunFor(
+                                            spec,
+                                            contractCreate(WIPE_CONTRACT).bytecode(WIPE_CONTRACT),
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java:145]: It seems you create WIPE_CONTRACT twice. Also bytecode is not needed anymore to be passed explicitly.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+                                            contractCall(
+                                                            WIPE_CONTRACT,
+                                                            "wipeNonFungibleToken",
+                                                            asAddress(vanillaTokenID.get()),
+                                                            asAddress(accountID.get()),
+                                                            serialNumbers)
+                                                    .payingWith(ACCOUNT)
+                                                    .via("wipeFungibleTxn")
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java:153]: Rename to `wipeNonFungibleTxn`

File: test-clients/src/main/resource/contract/solidity/hip-206/HederaTokenService.sol [hunk 16]
```diff
@@ -270,6 +270,31 @@ abstract contract HederaTokenService is HederaResponseCodes {
...
+            abi.encodeWithSelector(IHederaTokenService.wipeTokenAccount.selector, token, account, amount));
+        (responseCode) = success ? abi.decode(result, (int32)) : HederaResponseCodes.UNKNOWN;
+    }
+
+    /// Operation to wipe non fungible tokens from account
+    /// @param token The token address
+    /// @param account The account address to revoke kyc
+    /// @param  serialNumbers The serial numbers of token to wipe
+    /// @return responseCode The response code for the status of the request. SUCCESS is 22.
+    function wipeTokenAccount(address token, address account, int64[] memory serialNumbers) internal
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/resource/contract/solidity/hip-206/HederaTokenService.sol:290]: wipeTokenAccount -> wipeTokenAccountNFT

**Review by Nana-EC — DISMISSED** [review]

**Review by stoqnkpL — APPROVED** [review]
LGTM

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    @Override
+    protected Logger getResultsLogger() {
+        return log;
+    }
+
+    @Override
+    public List<HapiApiSpec> getSpecsInSuite() {
+        return allOf(List.of(wipeFungibleTokenScenarios(), wipeNonFungibleTokenScenarios()));
```
Comment (mustafauzunn) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/WipeTokenAccountPrecompileSuite.java:70]: Wouldn't it better to shorten this to:
return List.of(...)

**Review by mustafauzunn — COMMENTED** [review]
LGTM
